### ISSUE TITLE
DTLS inactivity monitor with multiple connections

### ISSUE THREAD

**Issue by jdbruijn** [plgd-dev/go-coap#209]
@jkralik 

I'm using the `DTLSListener` with multiple goroutines and the inactivity monitor. With the inactivity monitor I just close the connection in the `onInactive` function and have a timeout of let's say 30 seconds. I'd be expecting the inactivity monitor to work on a connection basis. So when I create connection every 5 seconds and just send one message on each connection, I'd expect connection 1. being inactive and closed about 30 seconds later. What I'm seeing however, is that the connections are never inactive as each new connection updates the same timeout of the inactivity monitor. This means, given the timeout is never reached, all connections stay open indefinitely and the resource utilisation (RAM and goroutines) just keeps climbing up and up.

I'm still looking into the code further to analyse why exactly this is happening and how I could change it to do a per-connection inactivity. Maybe this is due to my setup and the fact that I'm using a custom `DTLSListener` in order to create the goroutines for simultaneous DLTS handshakes, and is strongly based on https://github.com/plgd-dev/go-coap/blob/fe45892348e53edc6b7aba9d80c46379d347904c/net/dtlslistener.go#L35.

I'll let you know when I get further, any pointers on whether this is a bug, expected behaviour, or an obvious issue with my custom listener would be greatly appreciated.

<details><summary>Inactivity monitor code</summary>

```go
dtls.WithInactivityMonitor(inactivityTimeout,
  func(cc inactivity.ClientConn) {
    log.Println("inactive\n\n\n\n")
    cc.Close()
  }),
```

</details>

<details><summary>Accept loop code</summary>

```go
func (l *DTLSListener) acceptLoop() {
	defer l.wg.Done()
	for i := 0; i < 1000; i++ {
		go func() {
			for {
				conn, err := l.listener.Accept()
				prometheus.Instance.Prometheus.Register.ActiveConnectionsOld.Inc()
				id := randSeq(12)
				log.Printf(
					"DTLSListener.acceptLoop,Prometheus,ActiveConnections,++,%s\n", id)

				if err != nil {
					log.Printf("DTLSListener.acceptLoop,%s error %v\n", id, err)
					select {
					case <-l.doneCh:
						return
					case l.connCh <- connData{conn: conn, err: err}:
					}
				} else {
					prometheus.Instance.Prometheus.Register.ActiveConnections.Inc()
					select {
					case l.connCh <- connData{conn: conn, err: err}:
					case <-l.doneCh:
						return
					}
				}
			}
		}()
	}
}
```

</details>

**Comment by jdbruijn** [comment]
I'm still trying to fix this issue by creating an `inactivityMonitor` per connection. The patch below shows what I've got so far and sometimes it works perfectly as shown in the image below. Each individual connection is closed once it becomes inactive for the given period (30 s). The second image shows how the behaviour without this patch, which clearly shows that the number of active connections just keeps increasing when new connections come in before the last inactive period is reached.

![image](https://user-images.githubusercontent.com/62570005/109298338-b4fbd400-7833-11eb-936d-0132cba528aa.png)

![image](https://user-images.githubusercontent.com/62570005/109299086-d4dfc780-7834-11eb-9ebb-5da88a066df1.png)

The problem I'm currently facing, most certainly due to the implementation of the fix and my limited knowledge of Go and go-coap, is that the code sometimes has the following runtime error. I can't quite figure out why just yet and I'm not sure whether the approach I've taken to fix this issue is valid. The error occurs on a line that doesn't have anything to do with the inactivity monitor, so I'm still trying to figure out how these changes would impact that.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x8fb6be]

goroutine 1048 [running]:
github.com/plgd-dev/go-coap/v2/net/monitor/inactivity.(*InactivityMonitor).Notify(0xc000097d10)
        /home/jdebruijn/withthegrid/projects/dtls-termination-proxy/modules/go-coap/net/monitor/inactivity/inacti
vitymonitor.go:39 +0xbe
github.com/plgd-dev/go-coap/v2/udp/client.(*ClientConn).Process(0xc000922480, 0xc0008e6000, 0x29, 0x10000, 0x10000, 0x10000)
        /home/jdebruijn/withthegrid/projects/dtls-termination-proxy/modules/go-coap/udp/client/clientconn.go:551+0xe2
github.com/plgd-dev/go-coap/v2/dtls.(*Session).Run(0xc00012ba40, 0xc000922480, 0x0, 0x0)
        /home/jdebruijn/withthegrid/projects/dtls-termination-proxy/modules/go-coap/dtls/session.go:127 +0x18e
github.com/plgd-dev/go-coap/v2/udp/client.(*ClientConn).Run(...)
        /home/jdebruijn/withthegrid/projects/dtls-termination-proxy/modules/go-coap/udp/client/clientconn.go:450
github.com/plgd-dev/go-coap/v2/dtls.(*Server).Serve.func3(0xc00003cfa0, 0xc000922480, 0xc0000ba8f0)
        /home/jdebruijn/withthegrid/projects/dtls-termination-proxy/modules/go-coap/dtls/server.go:245 +0x66
created by github.com/plgd-dev/go-coap/v2/dtls.(*Server).Serve
        /home/jdebruijn/withthegrid/projects/dtls-termination-proxy/modules/go-coap/dtls/server.go:243 +0x375
```

<details><summary>Inactivity monitor per connection patch</summary>

```diff
diff --git a/dtls/server.go b/dtls/server.go
index 8527b44..59ef291 100644
--- a/dtls/server.go
+++ b/dtls/server.go
@@ -3,6 +3,7 @@ package dtls
 import (
 	"context"
 	"fmt"
+	"log"
 	"net"
 	"sync"
 	"time"
@@ -193,6 +194,7 @@ func (s *Server) checkAcceptError(err error) (bool, error) {
 }
 
 func (s *Server) Serve(l Listener) error {
+	log.Println("Serve!\n\n")
 	if s.blockwiseSZX > blockwise.SZX1024 {
 		return fmt.Errorf("invalid blockwiseSZX")
 	}
@@ -206,6 +208,13 @@ func (s *Server) Serve(l Listener) error {
 		s.listen = nil
 	}()
 
+	// inactivityMonitor := inactivity.NewInactivityMonitor(
+	// 	s.inactivityMonitor.Interval(),
+	// 	func(cc inactivity.ClientConn) {
+	// 		log.Println("inactivityMonitor => inactive")
+	// 		cc.Close()
+	// 	})
+
 	var wg sync.WaitGroup
 	defer wg.Wait()
 	for {
@@ -219,7 +228,14 @@ func (s *Server) Serve(l Listener) error {
 		}
 		if rw != nil {
 			wg.Add(1)
-			cc := s.createClientConn(coapNet.NewConn(rw, coapNet.WithHeartBeat(s.heartBeat)))
+
+			inactivityMonitor := inactivity.NewInactivityMonitorP(30*time.Second,
+				func(cc inactivity.ClientConn) {
+					log.Println("inactivityMonitor => inactive")
+					cc.Close()
+				})
+			cc := s.createClientConn(coapNet.NewConn(rw, coapNet.WithHeartBeat(s.heartBeat)),
+				inactivityMonitor)
 			if s.onNewClientConn != nil {
 				dtlsConn := rw.(*dtls.Conn)
 				s.onNewClientConn(cc, dtlsConn)
@@ -242,10 +258,15 @@ func (s *Server) Serve(l Listener) error {
 				}()
 			}
 			if s.inactivityMonitor != nil {
+				// if cc.ActivityMonitor != nil {
 				wg.Add(1)
 				go func() {
 					defer wg.Done()
-					err := s.inactivityMonitor.Run(cc)
+					log.Printf("Serve() => Call inactivityMonitor.Run for %s\n",
+						cc.RemoteAddr().String())
+					// err := s.inactivityMonitor.Run(cc, cc.RemoteAddr())
+					// err := cc.ActivityMonitor.Run(cc, cc.RemoteAddr())
+					err := inactivityMonitor.Run(cc, cc.RemoteAddr())
 					if err != nil {
 						s.errors(err)
 					}
@@ -260,7 +281,9 @@ func (s *Server) Stop() {
 	s.cancel()
 }
 
-func (s *Server) createClientConn(connection *coapNet.Conn) *client.ClientConn {
+func (s *Server) createClientConn(connection *coapNet.Conn,
+	inactivityMonitor *inactivity.InactivityMonitor) *client.ClientConn {
+	log.Printf("Server.createClientConn for %s\n", connection.RemoteAddr().String())
 	var blockWise *blockwise.BlockWise
 	if s.blockwiseEnable {
 		blockWise = blockwise.NewBlockWise(
@@ -281,6 +304,7 @@ func (s *Server) createClientConn(connection *coapNet.Conn) *client.ClientConn {
 		s.maxMessageSize,
 		true,
 	)
+
 	cc := client.NewClientConn(
 		session,
 		obsHandler,
@@ -294,7 +318,15 @@ func (s *Server) createClientConn(connection *coapNet.Conn) *client.ClientConn {
 		s.goPool,
 		s.errors,
 		s.getMID,
-		s.inactivityMonitor,
+		// s.inactivityMonitor,
+		inactivityMonitor,
+		// inactivity.NewInactivityMonitor(30*time.Second,
+		// 	func(cc inactivity.ClientConn) {
+		// 		log.Println("IM inactive")
+		// 		cc.Close()
+		// 	}),
+		// inactivity.NewInactivityMonitor(s.inactivityMonitor.Interval(),
+		// 	s.inactivityMonitor.OnInactive()),
 	)
 
 	return cc
diff --git a/net/monitor/inactivity/inactivitymonitor.go b/net/monitor/inactivity/inactivitymonitor.go
index 4bb991f..bf7049f 100644
--- a/net/monitor/inactivity/inactivitymonitor.go
+++ b/net/monitor/inactivity/inactivitymonitor.go
@@ -2,13 +2,18 @@ package inactivity
 
 import (
 	"context"
+	"log"
+	"net"
 	"sync/atomic"
 	"time"
+	// coapNet "github.com/plgd-dev/go-coap/v2/net"
 )
 
 type Monitor interface {
-	Run(cc ClientConn) error
+	Run(cc ClientConn, addr net.Addr) error
 	Notify()
+	Interval() time.Duration
+	OnInactive() OnInactiveFunc
 }
 
 type OnInactiveFunc func(cc ClientConn)
@@ -18,18 +23,33 @@ type ClientConn interface {
 	Close() error
 }
 
-type inactivityMonitor struct {
+type InactivityMonitor struct {
 	inactiveInterval time.Duration
 	onInactive       OnInactiveFunc
 	// lastActivity stores time.Time
 	lastActivity atomic.Value
+	addr         net.Addr
 }
 
-func (m *inactivityMonitor) Notify() {
+func (m *InactivityMonitor) Notify() {
 	m.lastActivity.Store(time.Now())
+
+	log.Printf("inactivityMonitor.Notify => Update for %s\n"+
+		"  last activity: %s => %s\n",
+		m.addr.String(),
+		m.LastActivity().UTC().Format(time.RFC3339),
+		time.Now().UTC().Format(time.RFC3339))
+}
+
+func (m *InactivityMonitor) Interval() time.Duration {
+	return m.inactiveInterval
 }
 
-func (m *inactivityMonitor) LastActivity() time.Time {
+func (m *InactivityMonitor) OnInactive() OnInactiveFunc {
+	return m.onInactive
+}
+
+func (m *InactivityMonitor) LastActivity() time.Time {
 	if t, ok := m.lastActivity.Load().(time.Time); ok {
 		return t
 	}
@@ -41,20 +61,35 @@ func CloseClientConn(cc ClientConn) {
 }
 
 func NewInactivityMonitor(interval time.Duration, onInactive OnInactiveFunc) Monitor {
-	return &inactivityMonitor{
+	log.Println("NewInactivityMonitor")
+	return &InactivityMonitor{
 		inactiveInterval: interval,
 		onInactive:       onInactive,
 	}
 }
 
-func (m *inactivityMonitor) Run(cc ClientConn) error {
+func NewInactivityMonitorP(interval time.Duration, onInactive OnInactiveFunc) *InactivityMonitor {
+	log.Println("NewInactivityMonitorP")
+	return &InactivityMonitor{
+		inactiveInterval: interval,
+		onInactive:       onInactive,
+	}
+}
+
+func (m *InactivityMonitor) Run(cc ClientConn, addr net.Addr) error {
 	if m.onInactive == nil || m.inactiveInterval == time.Duration(0) {
 		return nil
 	}
+	m.addr = addr
 
 	ticker := time.NewTicker(time.Second)
 	defer ticker.Stop()
 
+	log.Printf("inactivityMonitor.Run => Running for %s\n"+
+		"  last activity: %s\n",
+		m.addr.String(),
+		m.LastActivity().UTC().Format(time.RFC3339))
+
 	for {
 		timeout := time.Until(m.LastActivity().Add(m.inactiveInterval))
 		if timeout <= 0 {
@@ -63,6 +98,10 @@ func (m *inactivityMonitor) Run(cc ClientConn) error {
 		select {
 		case <-time.After(timeout):
 			if time.Since(m.LastActivity()) >= m.inactiveInterval {
+				log.Printf("inactivityMonitor.Run => %s has timed out\n"+
+					"  last activity: %s\n",
+					m.addr.String(),
+					m.LastActivity().UTC().Format(time.RFC3339))
 				m.onInactive(cc)
 			}
 		case <-cc.Context().Done():
diff --git a/udp/client/clientconn.go b/udp/client/clientconn.go
index e76a5aa..f6baf9c 100644
--- a/udp/client/clientconn.go
+++ b/udp/client/clientconn.go
@@ -5,6 +5,7 @@ import (
 	"crypto/sha256"
 	"fmt"
 	"io"
+	"log"
 	"net"
 	"strings"
 	"sync/atomic"
@@ -15,6 +16,7 @@ import (
 	"github.com/patrickmn/go-cache"
 	"github.com/plgd-dev/go-coap/v2/message"
 	"github.com/plgd-dev/go-coap/v2/net/blockwise"
+	"github.com/plgd-dev/go-coap/v2/net/monitor/inactivity"
 
 	"github.com/plgd-dev/go-coap/v2/message/codes"
 	udpMessage "github.com/plgd-dev/go-coap/v2/udp/message"
@@ -41,6 +43,7 @@ type Session interface {
 
 type Notifier interface {
 	Notify()
+	Run(cc inactivity.ClientConn, addr net.Addr) error
 }
 
 // ClientConn represents a virtual connection to a conceptual endpoint, to perform COAPs commands.
@@ -105,6 +108,7 @@ func NewClientConn(
 	getMID GetMIDFunc,
 	activityMonitor Notifier,
 ) *ClientConn {
+	log.Printf("NewClientConn for %s\n", session.RemoteAddr().String())
 	if errors == nil {
 		errors = func(error) {}
 	}
@@ -543,6 +547,7 @@ func (cc *ClientConn) getResponseFromCache(mid uint16, resp *pool.Message) (bool
 }
 
 func (cc *ClientConn) Process(datagram []byte) error {
+	log.Printf("ClientConn.Process for %s\n", cc.RemoteAddr().String())
 	cc.activityMonitor.Notify()
 	if cc.session.MaxMessageSize() >= 0 && len(datagram) > cc.session.MaxMessageSize() {
 		return fmt.Errorf("max message size(%v) was exceeded %v", cc.session.MaxMessageSize(), len(datagram))
diff --git a/udp/client/inactivitymonitor.go b/udp/client/inactivitymonitor.go
index f7f82c7..aac5000 100644
--- a/udp/client/inactivitymonitor.go
+++ b/udp/client/inactivitymonitor.go
@@ -1,8 +1,18 @@
 package client
 
+import (
+	"net"
+
+	"github.com/plgd-dev/go-coap/v2/net/monitor/inactivity"
+)
+
 // nilNotifier implements a Notify method that does nothing
 type nilNotifier struct {
 }
 
 func (n *nilNotifier) Notify() {
 }
+
+func (n *nilNotifier) Run(cc inactivity.ClientConn, addr net.Addr) error {
+	return nil
+}

```

</details>

**Comment by jdbruijn** [comment]
And in the original code when I stop creating new requests, all connections will be closed due to inactivity after the configured inactive period.

![image](https://user-images.githubusercontent.com/62570005/109299582-9dbde600-7835-11eb-89c1-fa2e0071a6d1.png)

**Comment by jkralik** [comment]
@jdbruijn I refactored inactivity monitor at master(it was not released yet). Could you try if it is fix the issue ?

**Comment by jdbruijn** [comment]
Thanks! I'll try with master in a bit and will let you know! Is that change supposed to make the inactivity monitor work on a per-connection basis?

**Comment by jkralik** [comment]
Yes it must works per connection. 

> 
> 
> Thanks! I'll try with master in a bit and will let you know! Is that change supposed to make the inactivity monitor work on a per-connection basis?

Yes It must works per connection.

**Comment by jdbruijn** [comment]
It works a treat now!

I'm using `dtls.WithKeepAlive(0, 0, nil),` and:

```go
dtls.WithInactivityMonitor(inactivityTimeout,
  func(cc inactivity.ClientConn) {
    log.Println("inactive\n\n\n\n")
    cc.Close()
  }),
```

Still with my custom DTLS listener to be able to process simultaneous DTLS handshakes. Any idea when there will be a new release from master and is the current master (fe45892) considered stable to use for production?

![image](https://user-images.githubusercontent.com/62570005/109303710-7833db00-783b-11eb-937c-b5429d657da8.png)

**Comment by jdbruijn** [comment]
@jkralik Also is `dtls.WithKeepAlive(0, 0, nil)` the correct way to disable keepalive ping-pongs?

**Comment by jkralik** [comment]
From https://github.com/plgd-dev/go-coap/releases/tag/v2.4.0 keepalive is disabled by default so you don't need to use it. You can directly use `dtls.WithInactivityMonitor` to close connection.

**Comment by jdbruijn** [comment]
Okay, I'll update my production to v2.4.0.

I consider this closed by 38a0437, and released in v2.4.0. Thanks for you help!
